### PR TITLE
Adding encode and decoding methods to Signals, Frames and CanMatrix

### PR DESCRIPTION
**Pull request by pierreluctg**
Adding methods that allows using canmatrix for generation and parsing of can frames

### CODE REVIEW

**Comment by ebroecker** [comment]
@pierreluctg:
Thanks for this code.
This feature has often been asked for.

From my point of view there is nothing against merging.
@altendky : you have some open questions, so should we wait with merging?

I'll look for some test-data. 
There is some in https://github.com/reinzor/libcan-encode-decode (which does the same but in C)...

**Comment by ebroecker** [comment]
sorry - didn't want to close ...

**Comment by pierreluctg** [comment]
Looks like there is an issue with new change on Python 3.x. working on a fix

File: canmatrix/canmatrix.py [hunk 1]
```diff
@@ -170,6 +175,31 @@ def normalizeValueTable(table):
     return {int(k): v for k, v in table.items()}
 
 
+class SignalValue:
+    """Proxy class to hold a Signal and value
+
+    Not using class "SignalValue(object):" on purpose to proxy magic methods
```
Comment (altendky) [inline canmatrix/canmatrix.py]: Are you saying you can't inherit from `object` (in other words, you have to use old style classes in py2) to specify your own `__getattr__`?  If so, [my quick check](https://repl.it/MExN) seems to work fine.

**Comment by altendky** [comment]
I've been using pytest in my own stuff.  It seems reasonable to settle on that if we were to switch to a 'real' testing framework rather than whatever I stuck in there before adding py3 support.

@ebroecker That's up to you, of course.  I would think it would be worth a little consideration of both the couple details I brought up and how it would be good to add this on.  But, this wouldn't break anything I have so it's not really a big deal.  As I mentioned, my code only uses canmatrix for parsing, basically.  Probably the most concrete thing to consider is using either `future` or `six` for 2/3 compatibility, but not both.

Presently, the encoding and decoding here are located on the signal and frame objects but if the latest received values are to be retained that is still an external activity.  Perhaps the signals and frames should have value, human value, and bytes sorts of properties to hold onto this?  Make a wrapper function for the decode that assigns back to the signals?  I see that the `SignalValue` class was added which creates some relationship but only one way and I'm not sure that hybridizing `Signal` with `int` or `float` etc really results in a sensible object.  I might be missing it but how do I get the value from a `SignalValue`?  `my_signal_value._value`?

The encode and decode probably need an option to ignore the min/max limits and a helper-function to just check that perhaps after decoding.  It should probably be a custom canmatrix exception class as well rather than ValueError so it can be caught independently.

I don't think I'm arguing against this code or merging, just trying to think things through beforehand.

**Comment by pierreluctg** [comment]
@altendky, thank you for the comments. 

You are correct, `SignalValue` is there to hold the actual value of a signal and allow proper `str` representation. I wanted to keep the signal definition `Signal` separated from the value.

Let say you hare a CAN sniffer that reads frames from the bus and prints the decoded data. Once you have parsed you DB, the `Signal` will be static. You only get various different `SignalValue`.

From a `Signal` class instance, you do not need access to `SignalValue`.

To get the value from a `SignalValue` you do not have to access `_value` or `value`, its a like a proxy class. I made few modification in the last commit to make it easier. So `my_signal_value` behave just like a int or a float (depends on signal.is_float).

**Comment by pierreluctg** [comment]
I have removed `six` now only using `future`

File: canmatrix/canmatrix.py (hunk 1 above)
Comment (pierreluctg) [reply canmatrix/canmatrix.py]: `__getattr__` works bu not for __eq__, __lt__, etc.

Comment (altendky) [reply canmatrix/canmatrix.py]: I was thinking on my lunch walk and agree that my example is irrelevant.  I'll try again.  But, if this truly only can be done on old-style classes how would it work in Python 3?

**Comment by altendky** [comment]
@pierreluctg Sure, if you only want the value at the time of decoding that's fine and really requires no class at all.  But, what if you want a place to lookup the most recent value at any given point in time?  I guess I just don't see much need for the pass-through class since we could instead just use regular methods for finding a signal object and then looking up the value in a dict keyed by the signal.  Or just keep it on the `Signal` itself.

Looking at the new approach, messing with `__new__` seems pretty questionable.  This seems like a lot to avoid adding one property and one method to Signal.  If separate, probably just have a `.signal` attribute on the new class to get at the signal when you want it.

File: canmatrix/canmatrix.py (hunk 1 above)
Comment (altendky) [reply canmatrix/canmatrix.py]: Alrighty, you shifted away from that, but, FYI, perhaps [`__getattribute__()`](https://docs.python.org/2/reference/datamodel.html#object.__getattribute__) was what you were missing?

File: tests/test_codec.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+                'Signal': u'two'
+            }
+            data_bytes = tuple(bytearray(bus.encode(test_frame1, data)))
+
+            decoded = bus.decode(test_frame1, data_bytes)
+
+            for k, v in data.items():
+                assert str(decoded[k]) == v
+
+    def test_encode_decode_signal_value_choice_str(self):
```
Comment (altendky) [inline tests/test_codec.py]: In py2, `str` is a byte sequence and, basically, should not be used.  It does wacky automatic stuff on top of just being confusing because of it's misleading name.

We probably want a single function and:
```
from __future__ import unicode_literals
from builtins import *
```
http://python-future.org/compatible_idioms.html#unicode-text-string-literals
http://python-future.org/compatible_idioms.html#basestring

If someone wants to encode a string to bytes, they need to pick their encoding which we can't do for them usefully.

I don't generally like the `*` in `from builtins import *` but I sort of feel like this is a special case where we want to make certain that as much as possible is made to be common.

Comment (pierreluctg) [reply tests/test_codec.py]: This is not encoding a string to bytes. `'two'` is a value description for the signal `Signal` in the frame `0x123`.

The value getting encoded will be `2`

File: canmatrix/canmatrix.py (hunk 1 above)
Comment (pierreluctg) [reply canmatrix/canmatrix.py]: Same issue with `__getattribute__` see the note in documentation.

File: tests/test_codec.py (hunk 2 above)
Comment (altendky) [reply tests/test_codec.py]: Yeah, pretty sure I misread this too...  again, I'm sorry.

Still, I think the imports are relevant for consistency of the environment the code runs in.  `str` should be a Unicode character sequence in keeping with elsewhere.  I may have messed up not importing `unicode_literals` elsewhere, I'm not sure.

Also, I don't think we want to be testing to make sure that people can pass byte sequences (py2 `str`) where we expect them to pass a Unicode sequence.  Sure, py2 does goofy stuff but it shouldn't work in py3.  Presently I'm thinking this would pass in py2 because of the automatic bytes->unicode conversion and in py3 because of `''` and `u''` being the same thing anyways.  With the imports I would expect both to pass in both py2 and py3 because they are identical at that point.  If you changed one to be `b''` then it would fail in both py2 and py3.  If I'm predicting correctly...

Comment (pierreluctg) [reply tests/test_codec.py]: I fully get your point now. SO you suggest removing `test_encode_decode_signal_value_choice_str`?

Comment (altendky) [reply tests/test_codec.py]: Along with adding the imports, sure.

File: canmatrix/formats.py [hunk 3]
```diff
@@ -43,18 +43,18 @@ def loads(string, importType=None, key="", flatImport=None, **options):
...
-            if path.endswith(extension) and "load" in supportedFormats[key]:
-                importType = key
-                break
-
-    if importType:
-        return load(fileObject, importType, **options)
-    else:
-        logger.error("This file format is not supported for reading")
-        return None
+    with open(path, "rb") as fileObject:
```
Comment (altendky) [inline canmatrix/formats.py]: Ack!  Yeah, much better...

**Comment by altendky** [comment]
I'm still a bit skeptical of `SignalValue`.  Messing with `__new__`, making a partially mutable `float`, etc.  The benefit just seems very slight for the funny business.

I believe the benefits in the present mechanism (`SignalValue` and the `OrderedDict` with signal name keys) are:

1. Decoded message can be indexed by the signal's name directly
2. Signal is accessible as well
3. Value is accessible
4. Automatically converts to a string

Perhaps we can still shift some functionality to a custom container such as a `DecodedFrame`?  Perhaps maintain a similar interface to the existing frames with `.signalByName()` and `.valueByName()` and `.strByName()` for quick by-name access.  One of those could be used in `__getitem__()`, though that might be better as a whole lib feature addition if it were to happen.  Or shift `.strByName()` onto a regular value class that just has the raw, human, and str attributes (or sure, `__str__()` `:]`).  If you want the raw and human values to be modifiable you could make them properties and recalculate the other based on an attribute referring to the original signal or cached copies of the factor/offset.

In my case I decided I didn't want my entire application to be tightly coupled to canmatrix so I made my own duplicative structure of frames and signals etc.  But, if this feature is going into canmatrix itself, why not just add the value and string conversion to the regular `Frame` and `Signal` classes?

**Comment by pierreluctg** [comment]
@altendky I agree that SignalValue is a bit messy. Will fix it.
